File: concurrent_scraper_py/utils/metrics.py
from dataclasses import dataclass, field
from typing import Dict
import time
from datetime import timedelta

@dataclass
class Metrics:
    start_time: float = None
    url_timings: Dict[str, float] = field(default_factory=dict)
    successful_requests: int = 0
    failed_requests: int = 0
    total_bytes: int = 0

    def start_operation(self):
        """Start timing an operation."""
        self.start_time = time.time()

    def record_success(self, url: str, duration: float, bytes_count: int):
        """Record a successful operation."""
        self.url_timings[url] = duration
        self.successful_requests += 1
        self.total_bytes += bytes_count

    def record_failure(self, url: str, duration: float):
        """Record a failed operation."""
        self.url_timings[url] = duration
        self.failed_requests += 1

    def get_summary(self) -> str:
        """Get a formatted summary of the metrics."""
        total_requests = self.successful_requests + self.failed_requests
        if total_requests == 0:
            return "No requests processed"

        avg_duration = sum(self.url_timings.values()) / total_requests

        return (
            f"Performance Summary:\n"
            f"Total Requests: {total_requests}\n"
            f"Successful: {self.successful_requests}\n"
            f"Failed: {self.failed_requests}\n"
            f"Average Duration: {timedelta(seconds=avg_duration)}\n"
            f"Total Data: {self.total_bytes} bytes"
        )


# For testing
if __name__ == "__main__":
    metrics = Metrics()
    metrics.start_operation()
    
    # Simulate some operations
    metrics.record_success("http://example.com", 1.5, 1000)
    metrics.record_failure("http://failed.com", 2.0)
    
    print(metrics.get_summary())

File: concurrent_scraper_rs/src/utils/metrics.rs
use std::time::{Duration, Instant};
use std::collections::HashMap;

#[derive(Debug, Default)]
pub struct Metrics {
    start_time: Option<Instant>,
    url_timings: HashMap<String, Duration>,
    successful_requests: usize,
    failed_requests: usize,
    total_bytes: usize,
}

impl Metrics {
    pub fn new() -> Self {
        Self {
            start_time: None,
            url_timings: HashMap::new(),
            successful_requests: 0,
            failed_requests: 0,
            total_bytes: 0,
        }
    }

    pub fn start_operation(&mut self) {
        self.start_time = Some(Instant::now());
    }

    pub fn record_success(&mut self, url: String, duration: Duration, bytes: usize) {
        self.url_timings.insert(url, duration);
        self.successful_requests += 1;
        self.total_bytes += bytes;
    }

    pub fn record_failure(&mut self, url: String, duration: Duration) {
        self.url_timings.insert(url, duration);
        self.failed_requests += 1;
    }

    pub fn get_summary(&self) -> String {
        let total_requests = self.successful_requests + self.failed_requests;
        let avg_duration = self.url_timings.values()
            .fold(Duration::default(), |acc, &x| acc + x)
            .div_f64(total_requests as f64);

        format!(
            "Performance Summary:\n\
             Total Requests: {}\n\
             Successful: {}\n\
             Failed: {}\n\
             Average Duration: {:.2?}\n\
             Total Data: {} bytes",
            total_requests,
            self.successful_requests,
            self.failed_requests,
            avg_duration,
            self.total_bytes
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_metrics_recording() {
        let mut metrics = Metrics::new();
        metrics.start_operation();
        
        metrics.record_success("http://example.com".to_string(), Duration::from_secs(1), 1000);
        metrics.record_failure("http://failed.com".to_string(), Duration::from_secs(2));

        assert_eq!(metrics.successful_requests, 1);
        assert_eq!(metrics.failed_requests, 1);
        assert_eq!(metrics.total_bytes, 1000);
    }
}
